### PR TITLE
Add accelerator method `teardown()`

### LINKED ISSUE #11091: Strategies and Accelerator teardown logic clean up
## Proposed refactor

After #11022 Move accelerator to strategies, the teardown logic can be simplified and cleaned up a bit. 

### Motivation

Reduce redundant code and make code more clean

### Pitch

In ddp, ddp_spawning and hovorod plugins, teardown() all has device specific logic like:
https://github.com/four4fish/pytorch-lightning/blob/move-accelerator/pytorch_lightning/plugins/training_type/ddp.py#L507-L511

with accelerator moved into strategy, device related teardown should move to X_accelerator
```
class GPUAccelerator()
        def teardown(self):
               # self.lightning_module.cpu()  suggested by @ananthsub module movement should be handled by strategies
               torch.cuda.empty_cache()
```

Some teardown() logic are shared between different strategy and same device type, like single_tpu and tpu_spawn has the same teardown() logic:
https://github.com/four4fish/pytorch-lightning/blob/move-accelerator/pytorch_lightning/plugins/training_type/single_tpu.py#L90-L92

we can move it to TPUAccelerator
```
class TPUAccelerator()
          def teardown(self) -> None:
                 os.environ.pop("PT_XLA_DEBUG", None)
```

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @justusschock @awaelchli @akihironitta @kaushikb11 @ananthsub

### PR DESCRIPTION
**Pull request by edward-io**
## What does this PR do?

Addresses #11091

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by ananthsub — APPROVED** [review]

**Review by ananthsub — COMMENTED** [review]
Please also add a unit test for each of the accelerator teardown implementations

**Comment by edward-io** [comment]
add tests - it doesn't seem like there's a good way to verify that cuda cache is emptied, so I'm only checking that teardown is called.

File: tests/accelerators/test_ddp.py [hunk 1]
```diff
@@ -145,3 +146,19 @@ def on_train_start(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule")
         callbacks=CustomCallback(),
     )
     trainer.fit(model)
+
+
+@RunIf(min_gpus=2, min_torch="1.8.1", standalone=True)
+def test_ddp_wrapper(tmpdir):
+    model = BoringModel()
+    strategy = DDPStrategy()
+    with patch.object(DDPStrategy, "teardown", wraps=strategy.teardown) as mock:
```
Comment (carmocca) [inline tests/accelerators/test_ddp.py:155]: This context manager should wrap just the `trainer.fit` call

File: tests/accelerators/test_ddp_spawn.py [hunk 2]
```diff
@@ -81,3 +84,19 @@ def test_ddp_all_dataloaders_passed_to_fit(tmpdir):
     )
     trainer.fit(model, train_dataloaders=model.train_dataloader(), val_dataloaders=model.val_dataloader())
     assert trainer.state.finished, "DDP doesn't work with dataloaders passed to fit()."
+
+
+@RunIf(min_gpus=2, standalone=True)
+def test_ddp_wrapper(tmpdir):
+    model = BoringModel()
+    strategy = DDPSpawnStrategy()
+    with patch.object(DDPSpawnStrategy, "teardown", wraps=strategy.teardown) as mock:
```
Comment (carmocca) [inline tests/accelerators/test_ddp_spawn.py:93]: Same comment

File: tests/accelerators/test_ddp.py [hunk 3]
```diff
@@ -145,3 +146,19 @@ def on_train_start(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule")
...
+        trainer = Trainer(
+            default_root_dir=tmpdir,
+            fast_dev_run=True,
+            strategy="ddp",
+            gpus=2,
```
Comment (carmocca) [inline tests/accelerators/test_ddp.py:160]: This test does not need 2 gpus

File: tests/accelerators/test_tpu.py [hunk 4]
```diff
@@ -332,3 +333,15 @@ def test_devices_auto_choice_tpu():
     trainer = Trainer(accelerator="auto", devices="auto")
     assert trainer.devices == 8
     assert trainer.tpu_cores == 8
+
+
+@RunIf(tpu=True)
+def test_teardown(tmpdir):
+    model = BoringModel()
+    strategy = TPUSpawnStrategy()
+    with patch.object(TPUSpawnStrategy, "teardown", wraps=strategy.teardown) as mock:
```
Comment (carmocca) [inline tests/accelerators/test_tpu.py:342]: Same

File: tests/accelerators/test_ddp_spawn.py [hunk 5]
```diff
@@ -81,3 +84,19 @@ def test_ddp_all_dataloaders_passed_to_fit(tmpdir):
...
+        trainer = Trainer(
+            default_root_dir=tmpdir,
+            fast_dev_run=True,
+            strategy="ddp",
+            gpus=2,
```
Comment (carmocca) [inline tests/accelerators/test_ddp_spawn.py:98]: Same

File: tests/accelerators/test_tpu.py [hunk 6]
```diff
@@ -332,3 +333,15 @@ def test_devices_auto_choice_tpu():
...
+        trainer = Trainer(strategy=strategy, max_epochs=1, tpu_cores=8, default_root_dir=tmpdir, fast_dev_run=True)
```
Comment (carmocca) [inline tests/accelerators/test_tpu.py:343]: redundant with fast_dev_run
```suggestion
        trainer = Trainer(strategy=strategy, tpu_cores=8, default_root_dir=tmpdir, fast_dev_run=True)
```

File: tests/accelerators/test_ddp.py [hunk 7]
```diff
@@ -145,3 +146,19 @@ def on_train_start(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule")
         callbacks=CustomCallback(),
     )
     trainer.fit(model)
+
+
+@RunIf(min_gpus=2, min_torch="1.8.1", standalone=True)
```
Comment (carmocca) [inline tests/accelerators/test_ddp.py:151]: This should work in 1.7 too

```suggestion
@RunIf(min_gpus=2, standalone=True)
```

File: pytorch_lightning/strategies/single_tpu.py [hunk 8]
```diff
@@ -83,5 +83,4 @@ def register_strategies(cls, strategy_registry: Dict) -> None:
 
     def teardown(self) -> None:
         super().teardown()
-        # TPU teardown
-        os.environ.pop("PT_XLA_DEBUG", None)
+        self.accelerator.teardown()
```
Comment (four4fish) [inline pytorch_lightning/strategies/single_tpu.py:86]: I think we can remove this override? The TPU tear down logic is in accelerator, single device and tpu single device shares the same teardown logic now

**Review by four4fish — APPROVED** [review]

**Review by rohitgr7 — APPROVED** [review]

File: tests/strategies/test_ddp_spawn.py [hunk 9]
```diff
@@ -81,3 +84,19 @@ def test_ddp_all_dataloaders_passed_to_fit(tmpdir):
     )
     trainer.fit(model, train_dataloaders=model.train_dataloader(), val_dataloaders=model.val_dataloader())
     assert trainer.state.finished, "DDP doesn't work with dataloaders passed to fit()."
+
+
+@RunIf(min_gpus=1, standalone=True)
+def test_ddp_wrapper(tmpdir):
```
Comment (akihironitta) [inline tests/strategies/test_ddp_spawn.py:90]:
```suggestion
def test_ddp_spawn_teardown(tmpdir):
```

File: tests/strategies/test_ddp.py [hunk 10]
```diff
@@ -145,3 +146,19 @@ def on_train_start(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule")
         callbacks=CustomCallback(),
     )
     trainer.fit(model)
+
+
+@RunIf(min_gpus=1, standalone=True)
+def test_ddp_wrapper_teardown(tmpdir):
```
Comment (akihironitta) [inline tests/strategies/test_ddp.py:152]:
```suggestion
def test_ddp_teardown(tmpdir):
```

**Review by akihironitta — APPROVED** [review]
nit

**Review by ananthsub — APPROVED** [review]

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/accelerators/tpu.py [hunk 11]
```diff
@@ -43,6 +44,9 @@ def get_device_stats(self, device: Union[str, torch.device]) -> Dict[str, Any]:
         }
         return device_stats
 
+    def teardown(self) -> None:
+        os.environ.pop("PT_XLA_DEBUG", None)
```
Comment (awaelchli) [inline pytorch_lightning/accelerators/tpu.py:48]: a teardown should not remove environment variables that it doesn't add itself.
This is done in the strategy, let's keep it there?

File: tests/strategies/test_ddp.py [hunk 12]
```diff
@@ -145,3 +146,19 @@ def on_train_start(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule")
         callbacks=CustomCallback(),
     )
     trainer.fit(model)
+
+
+@RunIf(min_gpus=1, standalone=True)
+def test_ddp_teardown(tmpdir):
```
Comment (awaelchli) [inline tests/strategies/test_ddp.py:152]: Thanks for thinking of a test! But in this case, let's do a unit test instead. 

We can't afford doing a full training run to check each individual hook call separately. We need to move away from this pattern. 

IMO there are two categories of tests

1) check that hooks get called and called in the right order
2) check that the implementation of a hook does what it should

To address 1) we don't need one test for each single hook. Instead, we can do what we have in `tests/models/test_hooks.py`
To address 2) we can instantiate the class and call the hook directly and assert the side-effects. [Random example](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/tests/plugins/precision/test_tpu_bf16_plugin.py)

**Review by awaelchli — CHANGES_REQUESTED** [review]

**Comment by edward-io** [comment]
Removed TPU teardown, replaced test with a specific one for GPU teardown.

File: pytorch_lightning/strategies/ddp.py [hunk 13]
```diff
@@ -429,5 +429,5 @@ def teardown(self) -> None:
             # GPU teardown
             log.detail(f"{self.__class__.__name__}: moving model to CPU")
             self.lightning_module.cpu()
-            # clean up memory
-            torch.cuda.empty_cache()
+
+        self.accelerator.teardown()
```
Comment (kaushikb11) [inline pytorch_lightning/strategies/ddp.py:433]: Should we add `self.accelerator.teardown()` to base class Strategy's `teardown` method as well?

File: pytorch_lightning/strategies/dp.py [hunk 14]
```diff
@@ -161,5 +161,5 @@ def teardown(self) -> None:
         if self.root_device.type == "cuda":
             # GPU teardown
             self.lightning_module.cpu()
```
Comment (carmocca) [inline pytorch_lightning/strategies/dp.py:163]: Do you plan to move this piece of code in a separate PR?

File: pytorch_lightning/accelerators/accelerator.py [hunk 15]
```diff
@@ -55,6 +55,9 @@ def get_device_stats(self, device: Union[str, torch.device]) -> Dict[str, Any]:
         """
         raise NotImplementedError
 
+    def teardown(self) -> None:
```
Comment (carmocca) [inline pytorch_lightning/accelerators/accelerator.py:58]: Remember to update the CHANGELOG with this addition

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/strategies/ddp.py (hunk 13 above)
Comment (four4fish) [reply pytorch_lightning/strategies/ddp.py:433]: I think this PR is reproducing the logic. Add self.accelerator.teardown() could change teardown logic for some strategies. Maybe let's try that out in a separate PR?

**Review by four4fish — APPROVED** [review]

File: pytorch_lightning/strategies/ddp.py (hunk 13 above)
Comment (awaelchli) [reply pytorch_lightning/strategies/ddp.py:433]: > Should we add self.accelerator.teardown() to base class Strategy's teardown method as well?

Adding it to the base makes sense to me. However, we need to be aware of the `super().teardown()` call. 
We have the lines  `self.lightning_module.cpu()` so the call to empty cuda cache should run after that.

I'm good with the current version. If you change the PR to what @kaushikb11 suggested, please re-request review. Otherwise LGTM

File: tests/accelerators/test_gpu.py [hunk 16]
```diff
@@ -72,3 +72,15 @@ def test_gpu_availability():
 def test_warning_if_gpus_not_used():
     with pytest.warns(UserWarning, match="GPU available but not used. Set `accelerator` and `devices`"):
         Trainer()
+
+
+@RunIf(min_gpus=1)
+def test_gpu_teardown():
+    acc = GPUAccelerator()
+    tensor = torch.zeros(100).cuda()
+    del tensor
```
Comment (awaelchli) [inline tests/accelerators/test_gpu.py:81]: are we counting on garbage collection being done at the appropriate time here?
It might be safer to add a `gc.collect()` call?

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/strategies/ddp.py (hunk 13 above)
Comment (tchaton) [reply pytorch_lightning/strategies/ddp.py:433]: Let's keep the PR atomic. One PR for replacing the code, one PR to move this in the base class.

File: pytorch_lightning/strategies/tpu_spawn.py [hunk 17]
```diff
@@ -287,6 +287,7 @@ def all_gather(self, tensor: torch.Tensor, group: Optional[Any] = None, sync_gra
     def teardown(self) -> None:
         super().teardown()
         os.environ.pop("PT_XLA_DEBUG", None)
```
Comment (tchaton) [inline pytorch_lightning/strategies/tpu_spawn.py:289]: Should this be moved to the TPU teardown accelerator @kaushikb11

**Review by tchaton — APPROVED** [review]
LGTM ! Small comments

File: pytorch_lightning/strategies/strategy.py [hunk 18]
```diff
@@ -442,7 +445,12 @@ def teardown(self) -> None:
         It is the right place to release memory and free other resources.
         """
         optimizers_to_device(self.optimizers, torch.device("cpu"))
+
+        if self.lightning_module is not None:
+            log.detail(f"{self.__class__.__name__}: moving model to CPU")
+            self.lightning_module.cpu()
```
Comment (carmocca) [inline pytorch_lightning/strategies/strategy.py:451]: Why not have this be part of the GPUAccelerator teardown?

This might be limiting for other accelerators (but that's an issue in current master too)

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/strategies/strategy.py (hunk 18 above)
Comment (justusschock) [reply pytorch_lightning/strategies/strategy.py:451]: AFAIK our design is to move models back to cpu in general after we are done, no? When we are supposed to release all device memory claimed, there is no other way then to move the model to cpu imo. And this should be handled by the strategy and not the accelerator since the strategy is owning the model :)
